backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/EnumSpecialDescriptorsFactory.kt

```diff
@@ -8,38 +8,146 @@ package org.jetbrains.kotlin.backend.konan
-                                val valuesField: IrField,
-                                val valuesGetter: IrSimpleFunction,
-                                val itemGetterSymbol: IrSimpleFunctionSymbol,
-                                val entriesMap: Map<Name, Int>)
+/**
+ * Common interface for both [InternalLoweredEnum] and [ExternalLoweredEnum]
+ * that allows to work with lowered enum regardless of its location.
+ */
+internal interface LoweredEnumAccess {
+    val valuesGetter: IrSimpleFunction
+    val itemGetterSymbol: IrSimpleFunctionSymbol
+    val entriesMap: Map<Name, Int>
+    fun getValuesField(startOffset: Int, endOffset: Int): IrExpression
+}
+
+/**
+ * Represents lowered enum from current module.
+ */
+internal data class InternalLoweredEnum(
+        val implObject: IrClass,
+        val valuesField: IrField,
+        val valuesGetterWrapper: IrSimpleFunction,
+        override val valuesGetter: IrSimpleFunction,
+        override val itemGetterSymbol: IrSimpleFunctionSymbol,
+        override val entriesMap: Map<Name, Int>
+) : LoweredEnumAccess {
+    private fun internalObjectGetter(startOffset: Int, endOffset: Int) =
+            IrGetObjectValueImpl(startOffset, endOffset,
+                    implObject.defaultType,
+                    implObject.symbol
+            )
+
+    override fun getValuesField(startOffset: Int, endOffset: Int): IrExpression = IrGetFieldImpl(
+            startOffset,
+            endOffset,
+            valuesField.symbol,
+            valuesField.type,
+            internalObjectGetter(startOffset, endOffset)
+    )
+}
+
+/**
+ * Represents lowered enum that's located in external module.
+ */
+internal data class ExternalLoweredEnum(
+        override val valuesGetter: IrSimpleFunction,
+        override val itemGetterSymbol: IrSimpleFunctionSymbol,
+        override val entriesMap: Map<Name, Int>
+) : LoweredEnumAccess {
+    override fun getValuesField(startOffset: Int, endOffset: Int): IrExpression =
+            IrCallImpl(startOffset, endOffset, valuesGetter.returnType, valuesGetter.symbol)
+}
 
 internal class EnumSpecialDeclarationsFactory(val context: Context) {
     private val symbols = context.ir.symbols
 
-    fun createLoweredEnum(enumClass: IrClass): LoweredEnum {
+    private fun enumEntriesMap(enumClass: IrClass): Map<Name, Int> =
+            enumClass.declarations
+                    .filterIsInstance<IrEnumEntry>()
+                    .sortedBy { it.name }
+                    .withIndex()
+                    .associate { it.value.name to it.index }
+                    .toMap()
+
+    private fun findItemGetterSymbol(): IrSimpleFunctionSymbol =
+            symbols.array.functions.single { it.descriptor.name == Name.identifier("get") }
+
+    private fun valuesArrayType(enumClass: IrClass): IrType =
+            symbols.array.typeWith(enumClass.defaultType)
+
+    // We can't move property getter to the top-level scope.
```

Comment:
Why is that, BTW?

Reply:
Accessors should have the same parent as the property. IR validator catches it. 
Also it would a really surprising behaviour for IR-processing code :)

Reply:
I meant to turn it from being a getter to being a usual function and probably removing the property also. But looks like it will be an overkill anyway.

Reply:
Ah, ok. Yes, it would make the PR needlessly big.